### ISSUE TITLE
`boundingbox` cleanup

### ISSUE THREAD

**Issue by ffreyer** [MakieOrg/Makie.jl#2881]
In short I think we should

1. Clarify what `data_limits` and `boundingbox` does
2. Include the extent of, for example, a scattered marker in boundingbox
3. Add a `space` option in `boundingbox`
4. Add per-point `boundingboxes` and other intermediate stages

---

Currently we have both `data_limits` and `boundingbox` which for most plots only differ by an application of the model matrix

https://github.com/MakieOrg/Makie.jl/blob/bc13edb84e293ea26d143cb17789c480b4400129/src/layouting/boundingbox.jl#L1-L8

which often times is just an identity transform. The exception to this is `text` which returns a pixel space boundingbox for the scattered string(s). `text` also includes the string boundingbox depending on markerspace in `data_limits`

https://github.com/MakieOrg/Makie.jl/blob/84a1cdf49d718f19b6fee36978c265fa1ad3ab03/src/layouting/data_limits.jl#L43-L53

which goes against my understanding of what `data_limits` is supposed to be.

I think `data_limits` should only care about the positional arguments passed to a plot, as it does in most cases now. `boundingbox` on the other hand should be a tight boundingbox around the full plot, including markersize, linewidth, etc. Including those will require some back and forth transformation, e.g. space -> markerspace, add markersize, markerspace -> space for scatter. (1, 2)

In general it would be good to be able to specify the output space of a boundingbox. That way we could unify text with the other bounding boxes, and have some methods that skip the final transformation. (3) 

Making different stages of the boundingbox computation available would also be useful to skip some of steps which might be unnecessary in some use cases. The pipeline here might be: (4)
1. Per position `rawboundingboxes(plot)` without the position argument applied (probably always in markerspace, as positions may influence the result in other spaces)
2. Per position `boundingboxes(plot, space = plot.markerspace)` with positions applied (transform positions from space -> markerspace here, user defined output space)
3. The combined `boundingbox(plot, space = plot.space)` which merges all the individual bounding boxes from the previous step

These steps don't make sense for every (primitive) plot, but I think it's quite obvious when a plot has them and when it doesn't. Though it might be a bit more difficult to generalize this to recipe. Maybe we could just use `boundingboxes` on the child plots that implement them, or simply make that function only work on the primitive types.

**Comment by ffreyer** [comment]
Oh yea, the `boundingbox(plot, index)` I mentioned in #2800 would also be a good to avoid having to calculate all bounding boxes when you just care about one or a few elements.

**Comment by SimonDanisch** [comment]
We should also cache boundingboxes, since they're somewhat expensive to calculate, and it's quite natural to query for them across different plotting calls/ functions...

**Comment by asinghvi17** [comment]
That sounds good.  Maybe each atomic (or plot object) could cache its own boundingbox?  Or maybe just know if it's outdated or not (based on changes in input data and transformation matrix)

**Comment by ffreyer** [comment]
#3002 made me trip over this: 
https://github.com/MakieOrg/Makie.jl/blob/a26e6a1dd5c462755335f61ad852d493c6490c44/src/layouting/data_limits.jl#L140-L142
We should fix this too.

I also think it would be good to finish up #2766 before this and differentiate what would be world space and model/local space in opengl. I.e the space post `trans_func`and `model`, and the space before any transformations. That would give you a clean way to specify which transformations should be applied.

**Comment by asinghvi17** [comment]
It seems like the transform is being applied twice somewhere, but I can't figure out where.  If we can resolve that, then bounding boxes would probably be a lot easier to manage..

**Comment by ffreyer** [comment]
Another thing to consider here is how to handle different input spaces with children. Let's say we have
```
parent with undefined space
- child1 with space = :pixel
- child2 with space = :data
```
If we call `boundingbox(parent, space = :data)` should that exclude child1? Should it convert? Both have their use (exclude for limits, convert for e.g. text bounding boxes), so I think both should be available. 

We could add a `convert = (spaces...)` kwarg for this, which contains all the spaces which should be converted into the target space. Probably with `convert = :all` as a shorthand.

**Comment by ffreyer** [comment]
Fixing #2245 should be part of this

**Comment by ffreyer** [comment]
#3112  too

**Comment by ffreyer** [comment]
From #3176 / #3179:

The default pipeline for data limits of recipes is currently `data_limits > iterate_transformed > point_iterator` where `point_iterator` then recursively combines input arguments of child plots. This means that you have to implement a custom `point_iterator` instead of `data_limits` if you want to tweak a recipes reported data_limits. Otherwise wrapping that recipe in another recipe will return the wrong limits.

TODO: Move the recursion to `data_limits`. Should reduce allocations as well.

A similar problem appears with h/vlines and h/vspan. In this case we want to ignore one dimension which currently happens via `autolimits` attributes. If these recipes are part of another recipe these attributes get ignored by Axis, and limits become unreasonable. Maybe we could handle this through `data_limits` instead, making the dimension in question NaN?

**Comment by jkrumbiegel** [comment]
Yes let's continue the discussion here rather than in that other issue. I thought a `Union{Nothing,` approach might be nicer for a plot to communicate its raw limits, as then the merging could be sure that this is intended and not a computing artifact of something going wrong. I get that we need NaN as "nothing" for something like Cairo line gaps, but in normal Julia code I don't like it too much.

**Comment by ffreyer** [comment]
I always shy away from unions because of performance but it probably doesn't matter much here. The majority of time is going to be spend generating limits, not evaluating the output.

Something like `Union{Nothing, Rect}` wouldn't be enough here, because sometimes you only want to invalidate one dimension. `Rect{Union{Nothing, Float32}` seems fishy to me since it probably breaks a bunch of functions that calculate things, like `maximum`. So maybe we should make out own type instead?

I could also see some use in having the "invalid" limits around. For vlines and the like we really have one part data space limts and one part relative space limits (at least in theory) which could both be useful to know. So maybe we should just flag them by space with some kind of unknown/invalid/special space for when they don't  fit in any of the normal boxes.

**Comment by jkrumbiegel** [comment]
I thought `Tuple{Union{Nothing,Tuple{Float32,Float32}},Union{Nothing,Tuple{Float32,Float32}}}` and then at the end make a rect out of it.

**Comment by jkrumbiegel** [comment]
And for the special spaces, I would think we'd make special bbox request functions?
